### PR TITLE
Add levenshtein distance and UTs

### PR DESCRIPTION
**Pull request by mingwli**
[wiki levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) (there's DP implementation pseudo code in the middle)  
#3322

### CODE REVIEW

File: src/docfx/lib/StringUtility.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            if (srcLength == 0)
+            {
+                return targetLength;
+            }
+            if (targetLength == 0)
+            {
+                return srcLength;
+            }
+
+            int[,] matrix = new int[srcLength + 1, targetLength + 1];
```
Comment (yufeih) [inline src/docfx/lib/StringUtility.cs]: can we reuse this buffer using [System.Buffers.ArrayPool<T>](https://github.com/dotnet/corefx/blob/master/src/System.Buffers/src/System/Buffers/ArrayPool.cs) ?

File: src/docfx/lib/StringUtility.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class StringUtility
```
Comment (yufeih) [inline src/docfx/lib/StringUtility.cs]: consider calling this `Levenshtein.GetLevenshteinDistance` and does not make it an extension method

File: src/docfx/lib/StringUtility.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                {
+                    int cost = src[i - 1] == target[j - 1] ? 0 : 1;
+                    matrix[i, j] = new int[]
+                    {
+                        matrix[i - 1, j] + 1,
+                        matrix[i, j - 1] + 1,
+                        matrix[i - 1, j - 1] + cost,
+                    }.Min();
+                }
+            }
```
Comment (superyyrrzz) [inline src/docfx/lib/StringUtility.cs]: What about use recursion? We can choose `matrix[i - 1, j - 1] + cost` for sure when `cost==0`. Thus, in best case (src=target), complexity can be reduced from quadratic to linear.

File: src/docfx/lib/StringUtility.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                matrix[0, j] = j;
+            }
+
+            for (int j = 1; j <= targetLength; j++)
+            {
+                for (int i = 1; i <= srcLength; i++)
+                {
+                    int cost = src[i - 1] == target[j - 1] ? 0 : 1;
+                    matrix[i, j] = new int[]
```
Comment (yufeih) [inline src/docfx/lib/StringUtility.cs]: don't alloc a new `int[]` here, you could use a tuple

**Pull request merged**